plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java

```diff
@@ -0,0 +1,116 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public abstract class BaseJdbcCastPushdownTest
+        extends AbstractTestQueryFramework
+{
+    protected CastDataTypeTest left;
+    protected CastDataTypeTest right;
+
+    protected abstract SqlExecutor onRemoteDatabase();
+
+    protected abstract void setupTable();
+
+    protected abstract List<CastTestCase> supportedCastTypePushdown();
+
+    protected abstract List<CastTestCase> unsupportedCastTypePushdown();
+
+    protected abstract List<CastTestCase> failCast();
+
+    @BeforeAll
+    public void setup()
+    {
+        setupTable();
+    }
+
+    @AfterAll
+    public void cleanup()
+    {
+        if (left != null) {
+            left.close();
+        }
+        if (right != null) {
+            right.close();
+        }
+    }
+
+    @Test
+    public void testProjectionPushdownWithCast()
+    {
+        for (CastTestCase testCase : supportedCastTypePushdown()) {
+            assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), left.tableName())))
+                    .isFullyPushedDown();
+        }
+
+        for (CastTestCase testCase : unsupportedCastTypePushdown()) {
+            assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), left.tableName())))
+                    .isNotFullyPushedDown(ProjectNode.class);
+        }
+    }
+
+    @Test
+    public void testCastFails()
+    {
+        for (CastTestCase testCase : failCast()) {
+            assertThatThrownBy(() -> getQueryRunner().execute("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), left.tableName())))
+                    .hasMessageMatching("(.*)Cannot cast (.*) to (.*)");
+        }
+    }
+
+    @Test
+    public void testJoinPushdownWithCast()
+    {
+        // combination of supported types
+        for (CastTestCase testCase : supportedCastTypePushdown()) {
+            assertJoinFullyPushedDown(left.tableName(), right.tableName(), "JOIN", "CAST(l.%s AS %s) = r.%s".formatted(testCase.sourceColumn(), testCase.castType(), testCase.targetColumn().orElseThrow()));
+        }
+
+        // combination of unsupported types
+        for (CastTestCase testCase : unsupportedCastTypePushdown()) {
+            assertJoinNotFullyPushedDown(left.tableName(), right.tableName(), "CAST(l.%s AS %s) = r.%s".formatted(testCase.sourceColumn(), testCase.castType(), testCase.targetColumn().orElseThrow()));
+        }
+    }
+
+    protected void assertJoinFullyPushedDown(String leftTable, String rightTable, String joinType, String joinCondition)
+    {
+        assertJoin(leftTable, rightTable, joinType, joinCondition)
+                .isFullyPushedDown();
+    }
+
+    protected void assertJoinNotFullyPushedDown(String leftTable, String rightTable, String joinCondition)
+    {
+        assertJoin(leftTable, rightTable, "JOIN", joinCondition)
```

Comment:
why `assertJoinFullyPushedDown` has `joinType` as parameter, and here it's hardcoded?

Reply:
This method will be removed after inlining.